File: solutions/python/reverse_words.py
import sys
import random
import string


def rand_string(length):
    ret = bytearray(length)
    for i in range(length):
        ch = random.randint(0, 62)
        if ch > 51:
            ret[i] = ord(' ')
        else:
            ret[i] = ord(string.ascii_letters[ch])
    return ret


# @include
# Assume s is a string encoded as bytearray.
def reverse_words(s):
    # First, reverse the whole string.
    s.reverse()

    def reverse_range(s, start, end):
        while start < end:
            s[start], s[end] = s[end], s[start]
            start, end = start + 1, end - 1

    start = 0
    while True:
        end = s.find(b' ', start)
        if end < 0:
            break
        # Reverses each word in the string.
        reverse_range(s, start, end - 1)
        start = end + 1
    # Reverses the last word.
    reverse_range(s, start, len(s) - 1)
# @exclude


# Pythonic solution, doesn't reverse in-place, may be used with strings
def reverse_words_pythonic(s):
    return ' '.join(reversed(s.split(' ')))


def check_answer(ori, s):
    reverse_words(s)
    assert ori == s


def main():
    for _ in range(1000):
        s = bytearray(' '.join(sys.argv[1:]), 'utf-8') if len(
            sys.argv) >= 2 else rand_string(random.randint(0, 100))
        original_str = s.copy()
        print(s.decode())
        reverse_words(s)
        print(s.decode())
        assert s.decode() == reverse_words_pythonic(original_str.decode())
        check_answer(original_str, s)


if __name__ == '__main__':
    main()

File: static/cpp/java/src/main/java/com/epi/ReverseWords.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.
package com.epi;

import java.util.Random;

import static com.epi.utils.Utils.find;
import static com.epi.utils.Utils.reverse;

public class ReverseWords {
  static String randString(int len) {
    Random gen = new Random();
    StringBuilder ret = new StringBuilder();
    while (len-- > 0) {
      int ch = gen.nextInt(53);
      if (ch == 52) {
        ret.append(' ');
      } else if (ch < 26) {
        ret.append(ch + 'a');
      } else {
        ret.append(ch - 26 + 'A');
      }
    }
    return ret.toString();
  }

  // @include
  public static String reverseWords(String input) {
    // Reverses the whole string first.
    char[] reversed = input.toCharArray();
    reverse(reversed, 0, input.length());

    int start = 0, end;
    while ((end = find(reversed, ' ', start)) != -1) {
      // Reverses each word in the string.
      reverse(reversed, start, end);
      start = end + 1;
    }
    // Reverses the last word.
    reverse(reversed, start, reversed.length);

    return new String(reversed);
  }
  // @exclude

  static void checkAnswer(String ori, String str) {
    String reversed = reverseWords(str);
    assert ori.equals(reversed);
  }

  public static void main(String[] args) {
    Random gen = new Random();
    for (int times = 0; times < 1000; ++times) {
      String str = "";
      if (args.length >= 1) {
        str += args[0];
        for (int i = 1; i < args.length; ++i) {
          str += ' ';
          str += args[i];
        }
      } else {
        str = randString(gen.nextInt(10000));
      }
      System.out.println(str);
      String reversed = reverseWords(str);
      System.out.println(reversed);
      checkAnswer(str, reversed);
    }
  }
}
